### PR TITLE
[#30083][prism] Factor out hold tracking to dedicated structures

### PR DESCRIPTION
**Pull request by lostluck**
Refactor out hold tracking to dedicated structures, so the logic is not replicated everywhere. Document the purpose of watermark holds within that.

This is yet another change to make the processing time change easier to read. In particular, event time watermark hold management is critical for correct behavior when waiting for processing time triggers.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @damondouglas

File: sdks/go/pkg/beam/runners/prism/internal/engine/holds.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+			if hold == h {
+				heap.Remove(&ht.heap, i)
+				break
+			}
+		}
+	} else if n > 0 {
+		ht.counts[hold] = n
+	} else {
+		panic(fmt.Sprintf("prism error: negative watermark hold count %v for time %v", n, hold))
+	}
```
Comment (damondouglas) [inline sdks/go/pkg/beam/runners/prism/internal/engine/holds.go:84]: Instead could we:
```
if ht.counts[hold] == 0 {
    panic(fmt.Sprintf("prism error: watermark hold count for time %v at 0", hold))
}

n := ht.counts[hold] - v

if n > 0 {
   ht.counts[hold] = n
   return
}

delete(ht.counts, hold)
for i, h := range ht.heap {
	if hold == h {
		heap.Remove(&ht.heap, i)
		break
	}
}

```

**Review by damondouglas — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/beam/runners/prism/internal/engine/holds.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+// holdHeap orders holds based on their timestamps
+// so we can always find the minimum timestamp of pending holds.
+type holdHeap []mtime.Time
+
+func (h holdHeap) Len() int           { return len(h) }
+func (h holdHeap) Less(i, j int) bool { return h[i] < h[j] }
+func (h holdHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+
+func (h *holdHeap) Push(x any) {
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/engine/holds.go:33]: `container/heap` is a pre-generics standard library package, so it takes in an `any` instead. To conform to the [heap interface](https://pkg.go.dev/container/heap#Interface) it needs to match the method types exactly.

File: sdks/go/pkg/beam/runners/prism/internal/engine/holds.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	// Push and Pop use pointer receivers because they modify the slice's length,
+	// not just its contents.
+	*h = append(*h, x.(mtime.Time))
+}
+
+func (h *holdHeap) Pop() any {
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/engine/holds.go:39]: Same answer.

File: sdks/go/pkg/beam/runners/prism/internal/engine/holds.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/engine/holds.go:84]: Sort of, but that wouldn't be the same.

1. That only assumes that the value is 0 before dropping. The current code is checking if it's negative after dropping. The current code catches additional errors.
2. The current code also reduces the number of lookups that occur, and sets that occur.

That said, I will reduce the if-else ladder somewhat.

**Comment by lostluck** [comment]
Thanks! PTAL

**Review by damondouglas — APPROVED** [review]
@lostluck

**Pull request merged**